### PR TITLE
VMware: add new module: vmware_host_ntp

### PR DESCRIPTION
**Pull request by Akasurde**
##### SUMMARY
This module manages NTP configuration for given ESXi server.

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME

lib/ansible/modules/cloud/vmware/vmware_host_ntp.py
test/integration/targets/vmware_host_ntp/aliases
test/integration/targets/vmware_host_ntp/tasks/main.yml

##### ANSIBLE VERSION

```
2.5devel
```

### CODE REVIEW

**Comment by Akasurde** [comment]
cc @pdellaert @tchernomax

File: lib/ansible/modules/cloud/vmware/vmware_host_ntp.py [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+    - NTP settings are applied to every ESXi host system in given cluster.
+    - If C(esxi_hostname) is not given, this parameter is required.
+  esxi_hostname:
+    description:
+    - ESXi hostname.
+    - NTP settings are applied to this ESXi host system.
+    - If C(cluster_name) is not given, this parameter is required.
+  ntp_server:
+    description:
+    - "IP or FQDN of NTP server/s."
```
Comment (pdellaert) [inline lib/ansible/modules/cloud/vmware/vmware_host_ntp.py]: Probably good to add how you configure multiple servers (as a list, or specify `For multiple servers, please look at the examples`)

Very minor: Maybe `ntp_servers` (plural) is a better name? Just to indicate that more can be specified?

**Review by pdellaert — COMMENTED** [review]
One general question: Should this module also be able to manage the NTP service (start/stop/restart and enable on boot)?

**Comment by Akasurde** [comment]
@pdellaert For service management, I am planning to have a separate module which will be something similar to `service` module but specific to vmware ESXi host. This module will have all of these facility.

File: lib/ansible/modules/cloud/vmware/vmware_host_ntp.py (hunk 1 above)
Comment (Akasurde) [reply lib/ansible/modules/cloud/vmware/vmware_host_ntp.py]: ok.

**Review by pdellaert — APPROVED** [review]
Looks great! and sounds good on the service module

File: lib/ansible/modules/cloud/vmware/vmware_host_ntp.py [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+    - "present: Add NTP server/s if it's absent else do nothing."
+    - "absent: Remove NTP server/s if it's present else do nothing."
+    default: present
+    choices: [ present, absent ]
+extends_documentation_fragment: vmware.documentation
+'''
+
+EXAMPLES = r'''
+- name: Set NTP setting for all ESXi Host in given Cluster
+  vmware_host_config_facts:
```
Comment (tchernomax) [inline lib/ansible/modules/cloud/vmware/vmware_host_ntp.py]: vmware_host_config_facts → vmware_host_ntp

File: lib/ansible/modules/cloud/vmware/vmware_host_ntp.py [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+    username: '{{ vcenter_username }}'
+    password: '{{ vcenter_password }}'
+    cluster_name: cluster_name
+    state: present
+    ntp_servers:
+        - 0.pool.ntp.org
+        - 1.pool.ntp.org
+
+- name: Set NTP setting for an ESXi Host
+  vmware_host_config_facts:
```
Comment (tchernomax) [inline lib/ansible/modules/cloud/vmware/vmware_host_ntp.py]: vmware_host_config_facts → vmware_host_ntp

File: lib/ansible/modules/cloud/vmware/vmware_host_ntp.py [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+    username: '{{ vcenter_username }}'
+    password: '{{ vcenter_password }}'
+    esxi_hostname: '{{ esxi_hostname }}'
+    state: present
+    ntp_servers:
+        - 0.pool.ntp.org
+        - 1.pool.ntp.org
+
+- name: Remove NTP setting for an ESXi Host
+  vmware_host_config_facts:
```
Comment (tchernomax) [inline lib/ansible/modules/cloud/vmware/vmware_host_ntp.py]: vmware_host_config_facts → vmware_host_ntp

File: lib/ansible/modules/cloud/vmware/vmware_host_ntp.py [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+
+        return changed
+
+    def check_host_state(self):
+        change_list = []
+        changed = False
+        for host in self.hosts:
+            state = 'absent'
+
+            host_ntp_servers = self.check_ntp_servers(host)
```
Comment (tchernomax) [inline lib/ansible/modules/cloud/vmware/vmware_host_ntp.py]: @Akasurde Maybe you should change "host_ntp_servers" for "update_ntp_list" or "update_ntp_servers" or "ntp_servers_to_change" ("ntp_servers_to_add_or_remove").
What do you think ?

File: lib/ansible/modules/cloud/vmware/vmware_host_ntp.py [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+            host_ntp_service = self.check_ntp_service(host)
```
Comment (tchernomax) [inline lib/ansible/modules/cloud/vmware/vmware_host_ntp.py]: @Akasurde Maybe you should change "host_ntp_service" for "is_host_ntp_running" or something more meaningful ?
What do you think ?

File: lib/ansible/modules/cloud/vmware/vmware_host_ntp.py [hunk 7]
```diff
@@ -0,0 +1,265 @@
...
+                    host_service_system.StopService(id='ntpd')
+                if service_state == 'restart':
+                    host_service_system.RestartService(id='ntpd')
+                changed = True
+            except (vim.fault.InvalidState, vim.fault.NotFound, vim.fault.HostConfigFault) as e:
+                self.results[host.name]['error'] = e.msg
+
+        return changed
+
+    def update_ntp_settings(self, host, ntp_servers):
```
Comment (tchernomax) [inline lib/ansible/modules/cloud/vmware/vmware_host_ntp.py]: @Akasurde Maybe you should rename "update_ntp_settings" to "add_ntp_servers".
What do you think ?

File: lib/ansible/modules/cloud/vmware/vmware_host_ntp.py [hunk 8]
```diff
@@ -0,0 +1,265 @@
...
+        change_list = []
+        changed = False
+        for host in self.hosts:
+            state = 'absent'
+
+            host_ntp_servers = self.check_ntp_servers(host)
+            host_ntp_service = self.check_ntp_service(host)
+
+            if host_ntp_servers and host_ntp_service:
+                state = 'present'
```
Comment (tchernomax) [inline lib/ansible/modules/cloud/vmware/vmware_host_ntp.py]: If self.desired_state is 'present' and all the ntp servers are already present (so host_ntp_servers == []), state take the value "absent"… so self.results[host.name]["current_state"] = "absent"… which is not what's expected.

File: lib/ansible/modules/cloud/vmware/vmware_host_ntp.py [hunk 9]
```diff
@@ -0,0 +1,265 @@
...
+                host_date_time_manager.UpdateDateTimeConfig(date_config_spec)
+                changed = True
+            except vim.fault.HostConfigFault as e:
+                self.results[host.name]['error'] = e.msg
+            except Exception as e:
+                self.results[host.name]['error'] = e
+
+        return changed
+
+    def delete_ntp_settings(self, host, ntp_servers):
```
Comment (tchernomax) [inline lib/ansible/modules/cloud/vmware/vmware_host_ntp.py]: @Akasurde Maybe you should rename "delete_ntp_settings" to "delete_ntp_servers".
What do you think ?

**Review by tchernomax — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/cloud/vmware/vmware_host_ntp.py (hunk 2 above)
Comment (Akasurde) [reply lib/ansible/modules/cloud/vmware/vmware_host_ntp.py]: Ok.

File: lib/ansible/modules/cloud/vmware/vmware_host_ntp.py (hunk 3 above)
Comment (Akasurde) [reply lib/ansible/modules/cloud/vmware/vmware_host_ntp.py]: OK.

File: lib/ansible/modules/cloud/vmware/vmware_host_ntp.py (hunk 4 above)
Comment (Akasurde) [reply lib/ansible/modules/cloud/vmware/vmware_host_ntp.py]: OK.

File: lib/ansible/modules/cloud/vmware/vmware_host_ntp.py (hunk 7 above)
Comment (Akasurde) [reply lib/ansible/modules/cloud/vmware/vmware_host_ntp.py]: OK.

File: lib/ansible/modules/cloud/vmware/vmware_host_ntp.py (hunk 9 above)
Comment (Akasurde) [reply lib/ansible/modules/cloud/vmware/vmware_host_ntp.py]: OK.

File: lib/ansible/modules/cloud/vmware/vmware_host_ntp.py (hunk 5 above)
Comment (Akasurde) [reply lib/ansible/modules/cloud/vmware/vmware_host_ntp.py]: OK.

File: lib/ansible/modules/cloud/vmware/vmware_host_ntp.py (hunk 6 above)
Comment (Akasurde) [reply lib/ansible/modules/cloud/vmware/vmware_host_ntp.py]: OK.

File: lib/ansible/modules/cloud/vmware/vmware_host_ntp.py (hunk 8 above)
Comment (Akasurde) [reply lib/ansible/modules/cloud/vmware/vmware_host_ntp.py]: Should I change `current_state` for host to `present` in that case?

File: lib/ansible/modules/cloud/vmware/vmware_host_ntp.py [hunk 10]
```diff
@@ -0,0 +1,215 @@
...
+            self.results[host.name].update(dict(
+                ntp_servers_to_change=ntp_servers_to_change,
+                desired_state=self.desired_state,
+            )
+            )
+
+            if not ntp_servers_to_change:
+                change_list.append(False)
+                self.results[host.name]['current_state'] = self.desired_state
+                continue
```
Comment (tchernomax) [inline lib/ansible/modules/cloud/vmware/vmware_host_ntp.py]: `continue` isn't necessary

File: lib/ansible/modules/cloud/vmware/vmware_host_ntp.py [hunk 11]
```diff
@@ -0,0 +1,215 @@
...
+            elif ntp_servers_to_change:
+                if self.desired_state == 'present':
+                    changed = self.update_ntp_settings(host=host, ntp_servers=ntp_servers_to_change)
```
Comment (tchernomax) [inline lib/ansible/modules/cloud/vmware/vmware_host_ntp.py]: update_ntp_settings → update_ntp_servers (or re-change the update_ntp_servers method name)

File: lib/ansible/modules/cloud/vmware/vmware_host_ntp.py [hunk 12]
```diff
@@ -0,0 +1,215 @@
...
+                    change_list.append(changed)
+                elif self.desired_state == 'absent':
+                    changed = self.update_ntp_settings(host=host, ntp_servers=ntp_servers_to_change, operation='delete')
```
Comment (tchernomax) [inline lib/ansible/modules/cloud/vmware/vmware_host_ntp.py]: update_ntp_settings → update_ntp_servers (or re-change the update_ntp_servers method name)

**Review by tchernomax — CHANGES_REQUESTED** [review]
Some minor mistake remain ; fix them and it will be good to me.

**Review by tchernomax — APPROVED** [review]
LGTM

**Comment by Akasurde** [comment]
@tchernomax @pdellaert Thanks for review.

**Pull request merged**